### DISCUSSION TITLE
How to retrieve values in list item fields

### DISCUSSION THREAD

**Discussion by sabrinadeveloper2023** [pnp/pnpcore#1212] [Q&A]
Hi,

I need to retrieve the value stored in the fields of specific item but based on the field type Datetime and text.

I am having difficulties understanding how to read the values.

I have read the documentation below

https://pnp.github.io/pnpcore/using-the-sdk/listitems-fields.html

but this explains how to assign the value based on the field type and I cannot figure it how to retrieve the value instead.

Please can you provide me an example on how to read the values for fields datetime and text? I will really appreciated.

This is an example of part of the code I am using:

var rootList = (await context.Web.Lists.GetByServerRelativeUrlAsync($"{context.Uri.PathAndQuery}{relativeURL}", p => p.RootFolder, p => p.Fields));

foreach (var field in rootList.Fields.AsRequested())
{
//Read value here
}

Thanks

**Comment by jansenbe** [comment]
See https://pnp.github.io/pnpcore/using-the-sdk/listitems-intro.html#reading-list-items

**Comment by sabrinadeveloper2023** [comment]
Thank you so much for your answer.

I am testing the below example from the documentation:

![image](https://github.com/pnp/pnpcore/assets/130606153/6cbfc862-b7d4-4ad8-b987-ebaab4ad558e)


When I inspect the fields in listItem, I only see 25 fields:

 
![image](https://github.com/pnp/pnpcore/assets/130606153/2819a159-c750-4e4d-9ace-b116bbc2721c)



 but below is my entire list of fields for that library, How I can add ALL fields to the listitem object?

Thanks

	"FieldsAndTitle": {
		"ContentTypeId": "Content Type ID",
		"_ModerationComments": "Approver Comments",
		"FileLeafRef": "Name",
		"Modified_x0020_By": "Document Modified By",
		"Created_x0020_By": "Document Created By",
		"File_x0020_Type": "File Type",
		"HTML_x0020_File_x0020_Type": "HTML File Type",
		"_SourceUrl": "Source URL",
		"_SharedFileIndex": "Shared File Index",
		"ComplianceAssetId": "Compliance Asset Id",
		"Title": "Title",
		"TemplateUrl": "Template Link",
		"xd_ProgID": "HTML File Link",
		"xd_Signature": "Is Signed",
		"_ShortcutUrl": "Shortcut URL",
		"_ShortcutSiteId": "Shortcut Site Id",
		"_ShortcutWebId": "Shortcut Web Id",
		"_ShortcutUniqueId": "Shortcut Unique Id",
		"_ExtendedDescription": "Description",
		"TriggerFlowInfo": "Trigger Flow Info",
		"MediaServiceMetadata": "MediaServiceMetadata",
		"MediaServiceFastMetadata": "MediaServiceFastMetadata",
		"AppSource": "AppSource",
		"Comments": "Comments",
		"Creator": "Creator",
		"IsRecord": "IsRecord",
		"Language": "Language",
		"Number": "Number",
		"_dlc_DocId": "Document ID Value",
		"_dlc_DocIdUrl": "Document ID",
		"_dlc_DocIdPersistId": "Persist ID",
		"MediaServiceImageTags": "Image Tags",
		"lcf76f155ced4ddcb4097134ff3c332f": "Image Tags_0",
		"TaxCatchAll": "Taxonomy Catch All Column",
		"MediaServiceDateTaken": "MediaServiceDateTaken",
		"MediaServiceOCR": "Extracted Text",
		"MediaServiceGenerationTime": "MediaServiceGenerationTime",
		"MediaServiceEventHashCode": "MediaServiceEventHashCode",
		"_ColorHex": "Color",
		"_ColorTag": "Color Tag",
		"_Emoji": "Emoji",
		"Phase": "Phase",
		"Subject": "Subject",
		"ObjectType": "ObjectType",
		"ID": "ID",
		"ContentType": "Content Type",
		"Created": "Created",
		"Author": "Created By",
		"Modified": "Modified",
		"Editor": "Modified By",
		"_HasCopyDestinations": "Has Copy Destinations",
		"_CopySource": "Copy Source",
		"_ModerationStatus": "Approval Status",
		"FileRef": "URL Path",
		"FileDirRef": "Path",
		"Last_x0020_Modified": "Modified",
		"Created_x0020_Date": "Created",
		"File_x0020_Size": "File Size",
		"FSObjType": "Item Type",
		"SortBehavior": "Sort Type",
		"PermMask": "Effective Permissions Mask",
		"PrincipalCount": "Principal Count",
		"CheckedOutUserId": "ID of the User who has the item Checked Out",
		"IsCheckedoutToLocal": "Is Checked out to local",
		"CheckoutUser": "Checked Out To",
		"UniqueId": "Unique Id",
		"ParentUniqueId": "Document Parent Identifier",
		"SyncClientId": "Client Id",
		"ProgId": "ProgId",
		"ScopeId": "ScopeId",
		"VirusStatus": "Virus Status",
		"CheckedOutTitle": "Checked Out To",
		"_CheckinComment": "Check In Comment",
		"LinkCheckedOutTitle": "Checked Out To",
		"_EditMenuTableStart": "Edit Menu Table Start",
		"_EditMenuTableStart2": "Edit Menu Table Start",
		"_EditMenuTableEnd": "Edit Menu Table End",
		"LinkFilenameNoMenu": "Name",
		"LinkFilename": "Name",
		"LinkFilename2": "Name",
		"DocIcon": "Type",
		"ServerUrl": "Server Relative URL",
		"EncodedAbsUrl": "Encoded Absolute URL",
		"BaseName": "Name",
		"FileSizeDisplay": "File Size",
		"MetaInfo": "Property Bag",
		"_Level": "Level",
		"_IsCurrentVersion": "Is Current Version",
		"ItemChildCount": "Item Child Count",
		"FolderChildCount": "Folder Child Count",
		"Restricted": "Restricted",
		"OriginatorId": "Originator Id",
		"NoExecute": "NoExecute",
		"ContentVersion": "Content Version",
		"_ComplianceFlags": "Label setting",
		"_ComplianceTag": "Retention label",
		"_ComplianceTagWrittenTime": "Retention label Applied",
		"_ComplianceTagUserId": "Label applied by",
		"_IsRecord": "Item is a Record",
		"BSN": "BLOB Sequence Number"
	}

**Reply by jansenbe** [reply]
@sabrinadeveloper2023 : most system fields are not returned using that approach, if you've defined a custom field it should. If you need those custom fields you can use one of the approaches that uses a CAML query and then specify the fields you want to return.

**Comment by sabrinadeveloper2023** [comment]
Understood, thanks for the clarification.

One more issue/question: I am uploading files to SP, updating some metadata fields, then sending the fields values to the source system in the response.

After I upload the document I am reloading the fields a follow:

await fileAdded.LoadAsync(p => p.ListItemAllFields);

Then I update several custom fields as follow:

fileAdded.ListItemAllFields["System"] = “some value”;
await fileAdded.ListItemAllFields.UpdateOverwriteVersionAsync();

One of the field I need to update is the Author. I have read the documentation on how to update the system properties and I found a couple of discussions where they indicate what kind of permission I need to update the system properties, so this is clear to me.

However, In that collection of fields, I see the field: AuthorId. I don’t see field Author.

 
![image](https://github.com/pnp/pnpcore/assets/130606153/698ac82f-fee5-4208-890b-d3b6d2ca9a95)


But in the example from the documentation, I should be able to see Author in the ListItemAllFields.

Would you please clarify this for me, what I am missing here?

Thanks

 
![image](https://github.com/pnp/pnpcore/assets/130606153/420559e0-a3d0-44f3-9fb5-278074d954ea)

**Reply by jansenbe** [reply]
@sabrinadeveloper2023 : just try the approach as documented, that should work fine. There is an Author field, it just was not by default retrieved.
